### PR TITLE
[staking] Implement Endorsement Statemanager

### PR DESCRIPTION
**Pull request by envestcc**
# Description

`EndorsementStatemanager` is used to manager Put and Get operations of endorsement

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/staking/endorsement_statemanager.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	return &EndorsementStateManager{StateManager: sm}
+}
+
+// Put puts the endorsement of a bucket
+func (esm *EndorsementStateManager) Put(bucketIndex uint64, endorse *Endorsement) error {
+	key := endorsementKey(bucketIndex)
+	if _, err := esm.PutState(endorse, protocol.NamespaceOption(_stakingNameSpace), protocol.KeyOption(key)); err != nil {
+		return err
+	}
+	return nil
```
Comment (CoderZhi) [inline action/protocol/staking/endorsement_statemanager.go:29]:
```
_, err := esm.PutState(endorse, protocol.NamespaceOption(_stakingNameSpace), protocol.KeyOption(endorsementKey(bucketIndex)))
return err
```

File: action/protocol/staking/endorsement_statemanager.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+func (esm *EndorsementStateManager) Get(bucketIndex uint64) (*Endorsement, error) {
+	key := endorsementKey(bucketIndex)
+	value := Endorsement{}
+	if _, err := esm.State(&value, protocol.NamespaceOption(_stakingNameSpace), protocol.KeyOption(key)); err != nil {
+		if errors.Is(err, state.ErrStateNotExist) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return &value, nil
```
Comment (CoderZhi) [inline action/protocol/staking/endorsement_statemanager.go:42]:
```
switch _, err := esm.State(&value, protocol.NamespaceOption(_stakingNameSpace), protocol.KeyOption(endorsementKey(bucketIndex))); errors.Cause(err) {
case nil:
    return &value, nil
case state.ErrStateNotExist:
    return nil, nil
default:
    return nil, err
}
```

File: action/protocol/staking/endorsement_statemanager.go (hunk 1 above)
Comment (envestcc) [reply action/protocol/staking/endorsement_statemanager.go]: updated

File: action/protocol/staking/endorsement_statemanager.go (hunk 2 above)
Comment (envestcc) [reply action/protocol/staking/endorsement_statemanager.go]: updated

File: action/protocol/staking/endorsement_statemanager.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	return err
+}
+
+// NewEndorsementStateReader creates a new endorsement state reader
+func NewEndorsementStateReader(sr protocol.StateReader) *EndorsementStateReader {
+	return &EndorsementStateReader{StateReader: sr}
+}
+
+// Get gets the endorsement of a bucket
+func (esm *EndorsementStateReader) Get(bucketIndex uint64) (*Endorsement, error) {
```
Comment (dustinxie) [inline action/protocol/staking/endorsement_statemanager.go:43]: `esr`

File: action/protocol/staking/endorsement_statemanager.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	value := Endorsement{}
+	switch _, err := esm.State(&value, protocol.NamespaceOption(_stakingNameSpace), protocol.KeyOption(endorsementKey(bucketIndex))); errors.Cause(err) {
+	case nil:
+		return &value, nil
+	case state.ErrStateNotExist:
+		return nil, nil
```
Comment (dustinxie) [inline action/protocol/staking/endorsement_statemanager.go:49]: let's retunr `state.ErrStateNotExist` to be consistent with current other codes

File: action/protocol/staking/endorsement_statemanager.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	default:
+		return nil, err
+	}
+}
+
+func endorsementKey(bucketIndex uint64) []byte {
+	key := []byte{_endorsement}
+	return append(key, byteutil.Uint64ToBytesBigEndian(bucketIndex)...)
+}
```
Comment (dustinxie) [inline action/protocol/staking/endorsement_statemanager.go:58]: add test for `Get, Put`

File: action/protocol/staking/endorsement_statemanager.go (hunk 3 above)
Comment (envestcc) [reply action/protocol/staking/endorsement_statemanager.go]: updated

File: action/protocol/staking/endorsement_statemanager.go (hunk 4 above)
Comment (envestcc) [reply action/protocol/staking/endorsement_statemanager.go]: updated

File: action/protocol/staking/endorsement_statemanager.go (hunk 5 above)
Comment (envestcc) [reply action/protocol/staking/endorsement_statemanager.go:51]: updated

File: action/protocol/staking/endorsement_statemanager_test.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+	// insert endorsement
+	bucketIndex := uint64(123)
+	endorse := &Endorsement{
+		ExpireHeight: 456,
+	}
+	err := esm.Put(bucketIndex, endorse)
+	r.NoError(err)
+	expect := Endorsement{}
+	_, err = esm.State(&expect, protocol.NamespaceOption(_stakingNameSpace), protocol.KeyOption(endorsementKey(bucketIndex)))
```
Comment (dustinxie) [inline action/protocol/staking/endorsement_statemanager_test.go:28]: this is basically `reader.Get()`, I think just combine these 2 tests

**Review by dustinxie — APPROVED** [review]

**Review by millken — APPROVED** [review]

**Pull request merged**